stablehlo/transforms/StablehloWrapInComposite.cpp

```diff
@@ -0,0 +1,219 @@
+// functionality of the provided `implOp`. The new function is named uniquely
+// and is set to private visibility.
+func::FuncOp buildFuncOpWrappingOperation(Operation* op, ModuleOp module) {
+  mlir::SymbolTable symbol_table(module);
+  // Create an OpBuilder, insertion point at the end of module's body.
+  mlir::OpBuilder builder(module);
+  builder.setInsertionPointToEnd(&module.getBodyRegion().back());
+
+  // Create the function operation, set private and add to the symbol table.
+  // SymbolTable will resolve all name conflicts.
+  Location loc = op->getLoc();
+  auto funcName = (op->getName().getStringRef() + ".impl").str();
+  mlir::func::FuncOp implFunc = builder.create<mlir::func::FuncOp>(
+      loc, funcName,
+      builder.getFunctionType(op->getOperandTypes(), op->getResultTypes()));
+  implFunc.setPrivate();
+  symbol_table.insert(implFunc);
+
+  Block* block = implFunc.addEntryBlock();
+  builder.setInsertionPointToEnd(block);
+  mlir::Operation* clonedOp = builder.clone(*op);
+  clonedOp->setOperands(block->getArguments());
+  builder.create<mlir::func::ReturnOp>(loc, clonedOp->getResults());
+
+  return implFunc;
+}
+
+// Returns true if the given operation should be wrapped in a CompositeOp.
+bool shouldWrapInComposite(Operation* op,
+                           const AttributePredicateMap& attributePredicateMap) {
+  auto it = attributePredicateMap.find(op->getName().getTypeID());
+  return it != attributePredicateMap.end() && it->second(op) != std::nullopt;
+}
+
+// A ConversionPattern that matches any operation and rewrites it as a
+// stablehlo::CompositeOp. The original operation's functionality is
+// encapsulated within a newly created private function.
+class ConvertGenericOp : public RewritePattern {
+ public:
+  explicit ConvertGenericOp(MLIRContext* context,
+                            AttributePredicateMap attributePredicateMap)
+      : RewritePattern(MatchAnyOpTypeTag(), /*benefit=*/1, context),
+        attributePredicateMap(attributePredicateMap) {}
+
+  LogicalResult matchAndRewrite(Operation* op,
+                                PatternRewriter& rewriter) const override {
+    if (!shouldWrapInComposite(op, attributePredicateMap)) {
+      return failure();
+    }
+
+    auto module = op->getParentOfType<ModuleOp>();
+    if (module == nullptr) {
+      return rewriter.notifyMatchFailure(op, "Failed to find enclosing module");
+    }
+
+    func::FuncOp decomposition = buildFuncOpWrappingOperation(op, module);
+    auto compositeName = op->getName().getStringRef();
+
+    auto attributePredicate =
+        attributePredicateMap.at(op->getName().getTypeID());
+    auto namedAttributes = attributePredicate(op);
+    auto compositeAttributes = rewriter.getDictionaryAttr(*namedAttributes);
+
+    auto compositeOp = rewriter.create<stablehlo::CompositeOp>(
+        op->getLoc(), op->getResultTypes(), op->getOperands(), compositeName,
+        compositeAttributes, decomposition.getSymName());
```

Comment:
We initialize the `compositeVersion` in the class, but not used here. Let's include this and add a test to make sure specified version appears.

Reply:
Good catch! Will fix it.